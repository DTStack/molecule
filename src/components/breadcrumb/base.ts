import { prefixClaName, getBEMElement } from 'mo/common/className';

export const defaultBreadcrumbClassName = prefixClaName('breadcrumb');
export const breadcrumbItemClassName = getBEMElement(
    defaultBreadcrumbClassName,
    'item'
);
export const breadcrumbLabelClassName = getBEMElement(
    defaultBreadcrumbClassName,
    'label'
);
