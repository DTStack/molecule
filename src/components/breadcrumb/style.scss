@import 'mo/style/common';

#{$breadcrumb} {
    display: flex;
    overflow: auto;
    white-space: nowrap;

    &__item {
        align-items: center;
        background-color: var(--breadcrumb-background);
        color: var(--breadcrumb-foreground);
        cursor: pointer;
        display: flex;
        justify-content: left;
        text-decoration: none;

        &:focus {
            color: var(--breadcrumb-focusForeground);
            outline: 1px solid var(--list-focusOutline);
            outline-offset: -1px;
        }

        &:hover {
            opacity: 0.8;
        }
    }

    &__label {
        align-items: center;
        background-position: center center;
        background-repeat: no-repeat;
        background-size: 16px;
        display: flex;
        height: 100%;
        justify-content: center;
        text-decoration: none;
        width: 100%;
    }
}
