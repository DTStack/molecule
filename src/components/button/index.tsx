import React, { forwardRef } from 'react';
import { classNames, getBEMModifier, prefixClaName } from 'mo/common/className';

type BtnSizeType = 'normal' | 'large';
export interface IButtonProps
    extends Omit<React.ComponentProps<'button'>, 'ref'> {
    disabled?: boolean;
    size?: BtnSizeType;
    onClick?(event: React.MouseEvent): void;
}

export const defaultButtonClassName = prefixClaName('btn');
export const normalButtonClassName = getBEMModifier(
    defaultButtonClassName,
    'normal'
);
export const largeButtonClassName = getBEMModifier(
    defaultButtonClassName,
    'large'
);
export const disableButtonClassName = getBEMModifier(
    defaultButtonClassName,
    'disabled'
);

export const Button = forwardRef(function (
    props: React.PropsWithChildren<IButtonProps>,
    ref: React.ForwardedRef<HTMLButtonElement> | undefined
) {
    const { className, children, size = 'normal', ...custom } = props;

    const disabled = props.disabled ? disableButtonClassName : null;

    const sizeClassName =
        size === 'large' ? largeButtonClassName : normalButtonClassName;

    const claNames = classNames(
        className,
        defaultButtonClassName,
        sizeClassName,
        disabled
    );

    return (
        <button ref={ref} className={claNames} {...custom}>
            {children}
        </button>
    );
});
