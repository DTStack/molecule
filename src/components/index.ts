export { ActionBar } from 'mo/components/actionBar';

export type {
    ActionBarItem,
    IActionBarProps,
    IActionBarItemProps,
} from 'mo/components/actionBar';

export { Breadcrumb } from './breadcrumb';

export type { IBreadcrumbProps, IBreadcrumbItemProps } from './breadcrumb';

export { Button } from './button';
export type { IButtonProps } from './button';

export { Checkbox } from './checkbox';
export type { ICheckboxProps } from './checkbox';

export { Collapse } from './collapse';
export type { ICollapseProps } from './collapse';

export { useContextMenu } from './contextMenu';
export type { IContextMenuProps } from './contextMenu';

export { useContextView } from './contextView';
export type { IContextViewProps, IContextView } from './contextView';

export { Modal } from './dialog';
export type { IModalProps, IModalFuncProps } from './dialog';

export { DropDown } from './dropdown';
export type { IDropDownProps } from './dropdown';

export { Icon } from './icon';
export type { IIconProps } from './icon';

export { Input } from './input';
export type { IInputProps, ITextAreaProps } from './input';

export { List, Item } from './list';
export type { IListProps, IItemProps } from './list';

export { Menu, MenuItem, SubMenu, MenuMode } from './menu';

export type { IMenuProps, IMenuItemProps, ISubMenuProps } from './menu';

export { MonacoEditor } from './monaco';
export type { IMonacoEditorProps } from './monaco';

export { Scrollable } from './scrollable';
export type { IScrollbarProps } from './scrollable';

export { Search } from './search';
export type { ISearchProps } from './search';

export { Select, Option } from './select';
export type { ISelectProps, ISelectOptionProps } from './select';

export { Tabs } from './tabs';
export type { ITabsProps, TabsType } from './tabs';

export { Tab } from './tabs/tab';
export type { ITabProps } from './tabs/tab';

export { Toolbar } from './toolbar';
export type { IToolbarProps } from './toolbar';

export { default as Tooltip } from './tooltip';
export type { IToolTipProps } from './tooltip';

export { default as TreeView } from './tree';
export type { ITreeProps, ITreeNodeItemProps } from './tree';

export { default as SplitPane } from './split';
export * from './split';

export { default as Display } from './display';
export type { IDisplayProps } from './display';
