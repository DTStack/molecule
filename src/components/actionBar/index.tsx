import React from 'react';
import { useCallback, useEffect, useRef } from 'react';
import {
    prefixClaName,
    classNames,
    getBEMElement,
    getBEMModifier,
} from 'mo/common/className';
import { useContextMenu } from 'mo/components/contextMenu';
import { IMenuItemProps, Menu } from 'mo/components/menu';
import { mergeFunctions } from 'mo/common/utils';
import Tooltip from '../tooltip';
import { Icon } from '../icon';

export interface IActionBarItemProps<T = any> {
    id?: string;
    name?: React.ReactNode;
    title?: string;
    icon?: string | JSX.Element;
    disabled?: boolean;
    checked?: boolean;
    data?: T;
    contextMenu?: IMenuItemProps[];
    className?: string;
    onContextMenuClick?: (
        e: React.MouseEvent,
        item: IMenuItemProps | undefined
    ) => void;
    onClick?(event: React.MouseEvent, item: IActionBarItemProps): void;
}

export interface IActionBarProps<T = any> {
    data: IActionBarItemProps<T>[];
    className?: string;
    onContextMenuClick?: (
        e: React.MouseEvent,
        item: IMenuItemProps | undefined
    ) => void;
    onClick?(event: React.MouseEvent, item: IActionBarItemProps): void;
}

const defaultActionBarClassName = prefixClaName('action-bar');
const containerClassName = getBEMElement(
    defaultActionBarClassName,
    'container'
);
const itemClassName = getBEMElement(defaultActionBarClassName, 'item');
const itemDisabledClassName = getBEMModifier(itemClassName, 'disabled');
const itemCheckedClassName = getBEMModifier(itemClassName, 'checked');

export function ActionBarItem(props: IActionBarItemProps) {
    const {
        id,
        title,
        name,
        data = {},
        contextMenu = [],
        onClick,
        icon,
        onContextMenuClick,
    } = props;
    const disabled = props.disabled ? itemDisabledClassName : null;
    const checked = props.checked ? itemCheckedClassName : null;
    const refItem = useRef(null);

    let contextViewMenu;

    const onClickMenuItem = useCallback(
        (e: React.MouseEvent, item: IMenuItemProps | undefined) => {
            onContextMenuClick?.(e, item);
            contextViewMenu?.dispose();
        },
        [contextMenu]
    );
    const renderContextMenu = () => (
        <Menu onClick={onClickMenuItem} data={contextMenu} />
    );

    useEffect(() => {
        if (contextMenu.length > 0) {
            contextViewMenu = useContextMenu({
                anchor: refItem.current,
                render: renderContextMenu,
            });
        }
        return function cleanup() {
            contextViewMenu?.dispose();
        };
    });

    const onClickItem = function (event) {
        if (props.disabled) {
            event.preventDefault();
            return;
        }
        if (onClick) {
            onClick(event, props);
        }
        if (contextMenu.length > -1 && contextViewMenu) {
            contextViewMenu.show({
                x: event.clientX,
                y: event.clientY,
            });
        }
    };

    return (
        <li
            id={id}
            ref={refItem}
            className={classNames(itemClassName, disabled, checked)}
            onClick={onClickItem}
            data-id={data.id}
        >
            <Tooltip overlay={<span>{title}</span>}>
                <Icon type={icon}>{name}</Icon>
            </Tooltip>
        </li>
    );
}

export function ActionBar<T = any>(props: IActionBarProps<T>) {
    const {
        data = [],
        onClick,
        onContextMenuClick,
        className,
        ...custom
    } = props;

    const claNames = classNames(defaultActionBarClassName, className);

    const items = data.map((item: IActionBarItemProps<T>, index) => (
        <ActionBarItem
            key={item.id}
            {...item}
            onContextMenuClick={onContextMenuClick}
            data-index={index}
            onClick={mergeFunctions(onClick, item.onClick)}
        />
    ));

    return (
        <div className={claNames} {...custom}>
            <ul className={containerClassName}>{items}</ul>
        </div>
    );
}
