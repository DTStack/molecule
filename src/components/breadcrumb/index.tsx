import React from 'react';
import { classNames } from 'mo/common/className';
import { Icon } from 'mo/components';
import type { UniqueId, HTMLElementProps } from 'mo/common/types';
import {
    defaultBreadcrumbClassName,
    breadcrumbItemClassName,
    breadcrumbLabelClassName,
} from './base';
import { getDataAttributionsFromProps } from 'mo/common/dom';

export interface IBreadcrumbItemProps extends HTMLElementProps {
    id: UniqueId;
    href?: string;
    name?: string;
    icon?: string | JSX.Element;
    render?(item: IBreadcrumbItemProps): React.ReactNode;

    [key: string]: any;
}

export interface IBreadcrumbProps extends HTMLElementProps {
    routes: IBreadcrumbItemProps[];
    separator?: React.ReactNode;
    onClick?(event: React.MouseEvent, item?: IBreadcrumbItemProps): void;

    [key: string]: any;
}

export function Breadcrumb(props: IBreadcrumbProps) {
    const {
        onClick,
        className,
        separator,
        routes = [],
        style,
        title,
        role,
        ...extra
    } = props;

    const claNames = classNames(defaultBreadcrumbClassName, className);
    const len = routes.length;
    const sep = separator || <Icon type="chevron-right" />;
    const dataAttrs = getDataAttributionsFromProps(extra);

    return (
        <div
            className={claNames}
            role={role}
            style={style}
            title={title}
            {...dataAttrs}
        >
            {routes.map((route, index) => {
                const {
                    id,
                    className,
                    title,
                    style,
                    href,
                    icon,
                    render,
                    name,
                    ...rest
                } = route;
                return (
                    <a
                        key={id}
                        className={classNames(
                            breadcrumbItemClassName,
                            className
                        )}
                        title={title}
                        style={style}
                        href={href}
                        tabIndex={0}
                        onContextMenu={(e) => e.preventDefault()}
                        onClick={(e) => onClick?.(e, route)}
                        {...getDataAttributionsFromProps(rest)}
                    >
                        <Icon type={icon} />
                        <span className={breadcrumbLabelClassName}>
                            {render ? render(route) : name}
                        </span>
                        {len - index > 1 ? sep : null}
                    </a>
                );
            })}
        </div>
    );
}
