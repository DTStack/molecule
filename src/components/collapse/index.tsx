import React, { useLayoutEffect, useRef, useState, useMemo } from 'react';
import { classNames } from 'mo/common/className';
import { HTMLElementProps, UniqueId } from 'mo/common/types';
import { getDataAttributionsFromProps } from 'mo/common/dom';

import {
    collapseActiveClassName,
    collapseContentClassName,
    collapseExtraClassName,
    collapseHeaderClassName,
    collapseItemClassName,
    collapsePaneClassName,
    collapseTitleClassName,
    collapsingClassName,
    defaultCollapseClassName,
} from './base';
import { IActionBarItemProps } from '../actionBar';
import { Icon } from '../icon';
import { Toolbar } from '../toolbar';
import SplitPane, { Pane } from '../split';

type RenderFunctionProps = (data: ICollapseItem) => React.ReactNode;
export interface ICollapseItem extends HTMLElementProps {
    id: UniqueId;
    name: string;
    hidden?: boolean;
    toolbar?: IActionBarItemProps[];
    renderPanel?: RenderFunctionProps;
    config?: {
        /**
         * Specify how much of the remaining space should be assigned to the item, default is 1
         *
         * It unfolds in its own content height or the `MAX_GROW_HEIGHT` rather than in calculated height
         */
        grow?: number;
    };

    [key: string]: any;
}

export interface ICollapseProps extends HTMLElementProps {
    activePanelKeys?: UniqueId[];
    data?: ICollapseItem[];
    onCollapseChange?: (keys: UniqueId[]) => void;
    onResize?: (resizes: number[]) => void;
    onToolbarClick?: (
        item: IActionBarItemProps,
        parentPanel: ICollapseItem
    ) => void;

    [key: string]: any;
}

/**
 * It's the max height for the item which set the grow to 0
 */
export const MAX_GROW_HEIGHT = 220;
// default collapse height, only contains header
export const HEADER_HEIGTH = 26;

export function Collapse({
    data = [],
    activePanelKeys: controlActivePanelKeys,
    className,
    title,
    style,
    role,
    onCollapseChange,
    onToolbarClick,
    onResize,
    ...restProps
}: ICollapseProps) {
    const [activePanelKeys, setActivePanelKeys] = useState<UniqueId[]>(
        new Array(data.length)
    );
    const [collapsing, setCollapsing] = useState(false);
    const wrapper = useRef<HTMLDivElement>(null);
    const [sizes, setSizes] = useState<number[]>(
        data.map((pane) => (pane.hidden ? 0 : HEADER_HEIGTH))
    );
    // cache the adjusted size for restoring the adjusted size in next uncollapsing
    const adjustedSize = useRef<number[]>([]);
    const first = useRef(true);

    const isUndefined = (key): boolean => {
        return key === undefined;
    };

    // compare two sizes to find the change one
    const compareTheSizes = (sizes: number[], otherSizes: number[]) => {
        for (let index = 0; index < sizes.length; index++) {
            if (sizes[index] !== otherSizes[index]) {
                return index + 1;
            }
        }
        return -1;
    };

    const handleSplitChange = (nextSizes: number[]) => {
        const index = compareTheSizes(sizes, nextSizes);
        if (index === -1) {
            return;
        }
        adjustedSize.current[index] = nextSizes[index];
        onResize?.(nextSizes);
        setSizes(nextSizes);
    };

    const handleToolbarClick = (
        e: React.MouseEvent,
        item: IActionBarItemProps,
        panel: ICollapseItem
    ) => {
        e.stopPropagation();
        onToolbarClick?.(item, panel);
    };

    const renderPanels = (
        data: ICollapseItem,
        render?: RenderFunctionProps
    ) => {
        if (render) {
            return render(data);
        }
        return null;
    };

    const handleChangeCallback = (key: UniqueId, index) => {
        const currentKeys = [...activePanelKeys];
        if (!isUndefined(currentKeys[index])) {
            delete currentKeys[index];
        } else {
            currentKeys[index] = key;
        }
        onCollapseChange?.(currentKeys);
        setActivePanelKeys(currentKeys.concat());
    };

    // perform smoothly the task to recalculate sizes
    const performSmoothSizes = () => {
        setCollapsing(true);
        performSizes();
        setTimeout(() => {
            setCollapsing(false);
        }, 300);
    };

    // perform the tasks to recalculate sizes
    const performSizes = () => {
        const activeLength = activePanelKeys.filter(
            (v) => !isUndefined(v)
        ).length;
        if (activeLength) {
            const { height } = wrapper.current!.getBoundingClientRect();
            let restHeight = height;
            let count = 0;
            // don't care of what the previous sizes are, the next sizes only contains:
            // 1. directly assignment for the next size is collapsing
            // 2. recalculate for the next size is uncollapsing
            const wipSizes = data.map((pane, index) => {
                const isHidden = pane.hidden;
                if (isHidden) {
                    return 0;
                }
                const willCollapsing = activePanelKeys.includes(pane.id);
                if (!willCollapsing) {
                    restHeight = restHeight - HEADER_HEIGTH;
                    return HEADER_HEIGTH;
                }

                // to get the height of content while grow is 0
                if (pane.config?.grow === 0) {
                    const correspondDOM = wrapper.current
                        ?.querySelector(
                            `.${collapseContentClassName}[data-collapse-index='${index}']`
                        )
                        ?.querySelector(`[data-content='${pane.id}']`);
                    if (!correspondDOM) {
                        restHeight = restHeight - HEADER_HEIGTH;
                        return HEADER_HEIGTH;
                    }
                    const { height: contentHeight } =
                        correspondDOM.getBoundingClientRect();
                    // for preventing the loss of DOM height, don't set the display to be none for DOM
                    const height = contentHeight + HEADER_HEIGTH;

                    if (height > MAX_GROW_HEIGHT) {
                        restHeight = restHeight - MAX_GROW_HEIGHT;
                        return MAX_GROW_HEIGHT;
                    }
                    restHeight = restHeight - height;
                    return height;
                }

                // there is a cached size
                if (typeof adjustedSize.current[index] !== 'undefined') {
                    restHeight = restHeight - adjustedSize.current[index];
                    return adjustedSize.current[index];
                }

                // count the sum of grow that isn't 0, for how many parts the remaing part should be divided into
                count = count + (pane.config?.grow || 1);
                // auto is a placeholder for calculation in next process
                return 'auto';
            });

            // count the average size for each auto
            const averageHeight = restHeight / count;
            const nextSizes = wipSizes.map((size, index) =>
                size === 'auto'
                    ? averageHeight * (data[index].config?.grow || 1)
                    : size
            );

            onResize?.(nextSizes);
            setSizes(nextSizes);
        } else {
            const nextSizes = data.map((pane) =>
                pane.hidden ? 0 : HEADER_HEIGTH
            );
            onResize?.(nextSizes);
            setSizes(nextSizes);
        }
    };

    const { nextSashes, allowResize } = useMemo(
        function () {
            const nextSashes: boolean[] = [];
            for (let i = 1; i < activePanelKeys.length; i++) {
                const prevPaneActive: boolean = !isUndefined(
                    activePanelKeys[i - 1]
                );
                const prePaneAutoHeight: boolean =
                    data[i - 1]?.config?.grow !== 0;
                const curPaneActive: boolean = !isUndefined(activePanelKeys[i]);
                const showSash =
                    prevPaneActive && prePaneAutoHeight && curPaneActive;

                nextSashes.push(showSash);
            }
            return {
                nextSashes,
                allowResize: data.map((pane, index) => {
                    if (pane?.config?.grow === 0) {
                        return false;
                    }
                    return !isUndefined(activePanelKeys[index]);
                }),
            };
        },
        [activePanelKeys, data]
    );

    useLayoutEffect(() => {
        if (!first.current) {
            performSmoothSizes();
        }

        first.current = false;
    }, [activePanelKeys, data]);

    useLayoutEffect(() => {
        Array.isArray(controlActivePanelKeys) &&
            setActivePanelKeys(controlActivePanelKeys);
    }, [controlActivePanelKeys]);
    const dataAttrProps = getDataAttributionsFromProps(restProps);

    return (
        <div
            ref={wrapper}
            className={classNames(defaultCollapseClassName, className)}
            title={title}
            style={style}
            role={role}
            {...dataAttrProps}
        >
            <SplitPane
                sizes={sizes}
                onChange={handleSplitChange}
                split="horizontal"
                allowResize={allowResize}
                showSashes={nextSashes}
                paneClassName={classNames(
                    collapsePaneClassName,
                    collapsing && collapsingClassName
                )}
            >
                {data.map((panel, index) => {
                    const isActive = activePanelKeys.includes(panel.id);
                    return (
                        <Pane key={panel.id} minSize={HEADER_HEIGTH}>
                            <div
                                className={classNames(
                                    panel.className,
                                    collapseItemClassName,
                                    isActive && collapseActiveClassName
                                )}
                                data-collapse-id={panel.id}
                            >
                                <div
                                    className={collapseHeaderClassName}
                                    tabIndex={0}
                                    onClick={() =>
                                        handleChangeCallback(panel.id, index)
                                    }
                                >
                                    <Icon
                                        type={
                                            isActive
                                                ? 'chevron-down'
                                                : 'chevron-right'
                                        }
                                    />
                                    <span className={collapseTitleClassName}>
                                        {panel.name}
                                    </span>
                                    <div className={collapseExtraClassName}>
                                        {isActive && (
                                            <Toolbar
                                                key={panel.id}
                                                data={panel.toolbar || []}
                                                onClick={(e, item) =>
                                                    handleToolbarClick(
                                                        e,
                                                        item,
                                                        panel
                                                    )
                                                }
                                            />
                                        )}
                                    </div>
                                </div>
                                <div
                                    className={collapseContentClassName}
                                    tabIndex={0}
                                    data-collapse-index={index}
                                >
                                    {renderPanels(panel, panel.renderPanel)}
                                </div>
                            </div>
                        </Pane>
                    );
                })}
            </SplitPane>
        </div>
    );
}
