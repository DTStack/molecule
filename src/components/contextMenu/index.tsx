import React from 'react';
import { HTMLElementType } from 'mo/common/dom';
import { IContextView, useContextView } from 'mo/components/contextView';

export interface IContextMenuProps {
    anchor: HTMLElementType;
    render: () => React.ReactNode;
}

export interface IContextMenu extends IContextView {}

export function useContextMenu(
    props: IContextMenuProps
): IContextMenu | undefined {
    const { anchor, render } = props;

    if (!anchor) {
        return undefined;
    }

    const contextView = useContextView({
        render,
    });

    const onContextMenu = (e: MouseEvent) => {
        e.preventDefault();
        contextView!.show(
            {
                x: e.clientX,
                y: e.clientY,
            },
            render
        );
    };

    anchor.addEventListener('contextmenu', onContextMenu);

    const dispose = () => {
        contextView!.hide();
        anchor.removeEventListener('contextmenu', onContextMenu);
    };

    return { ...contextView, dispose };
}
