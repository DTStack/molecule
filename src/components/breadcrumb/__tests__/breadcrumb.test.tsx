import React from 'react';
import renderer from 'react-test-renderer';
import { createEvent, fireEvent, render } from '@testing-library/react';
import { Breadcrumb } from 'mo/components/breadcrumb';
import type { IBreadcrumbItemProps } from 'mo/components/breadcrumb';
import {
    breadcrumbItemClassName,
    breadcrumbLabelClassName,
    defaultBreadcrumbClassName,
} from '../base';

const mockData: IBreadcrumbItemProps[] = new Array(3)
    .fill(1)
    .map((_, index) => ({
        id: index.toString(),
        name: `name${index}`,
    }));

describe('Test Breadcrumb Component', () => {
    test('Should render Breadcrumb multiply levels', () => {
        const component = renderer.create(<Breadcrumb routes={mockData} />);

        const breadcrumb = component.toJSON();
        expect(breadcrumb).toMatchSnapshot();
    });

    test('Should have multiply classNames', () => {
        const wrapper = render(
            <Breadcrumb routes={mockData} className="test" />
        );

        const container = wrapper.container.querySelector(
            `.${defaultBreadcrumbClassName}`
        ) as HTMLDivElement;

        expect(container.classList).toContain('test');
    });

    test('Should render correct hrefs in Breadcrumb', () => {
        const data = mockData.map((item, index) => ({
            ...item,
            className: 'test',
            href: `https://www.dtstack.com/#${index}`,
        }));
        const wrapper = render(<Breadcrumb routes={data} />);
        const doms = wrapper.container.querySelectorAll('.test');
        const result = Array.prototype.every.call(
            doms,
            (child: HTMLAnchorElement, index) => {
                return child.href === `https://www.dtstack.com/#${index}`;
            }
        );

        expect(result).toBeTruthy();
    });

    test('Should render icons in front of name in Breadcrumb', () => {
        const data = mockData.map((item) => ({
            ...item,
            className: 'test',
            icon: 'account',
        }));

        const wrapper = render(<Breadcrumb routes={data} />);
        const doms = wrapper.container.querySelectorAll('.test');
        doms.forEach((dom) => {
            const iconDom = dom.firstChild as HTMLSpanElement;
            const nameDom = iconDom.nextSibling as HTMLSpanElement;
            expect(iconDom.classList).toContain('codicon-account');
            expect(nameDom.classList).toContain(breadcrumbLabelClassName);
        });
    });

    test('Should support to custom-render', () => {
        const data = mockData.map((item) => ({
            ...item,
            className: 'test',
        }));
        data[0].render = () => <div id="customDiv">custom render div</div>;

        const wrapper = render(<Breadcrumb routes={data} />);
        const doms = wrapper.container.querySelectorAll('.test');
        const customDom = doms[0].querySelector('#customDiv');
        expect(customDom).not.toBeUndefined();
    });

    test('Should support modify separator & correct count of separators', () => {
        const wrapper = render(
            <Breadcrumb
                routes={mockData}
                separator={<span className="separator">!</span>}
            />
        );
        const separators = wrapper.container.querySelectorAll('.separator');

        const result = Array.prototype.every.call(
            separators,
            (child: HTMLSpanElement, index) => {
                return child.firstChild?.nodeValue === '!';
            }
        );

        expect(separators.length).toBe(2);
        expect(result).toBeTruthy();
    });

    test('Should trigger onClick', () => {
        const data = mockData.map((item) => ({
            ...item,
            className: 'test',
        }));

        const mockFn = jest.fn();
        const wrapper = render(<Breadcrumb routes={data} onClick={mockFn} />);
        const doms = wrapper.container.querySelectorAll('.test');
        doms.forEach((dom) => {
            fireEvent.click(dom);
        });
        expect(mockFn).toBeCalledTimes(data.length);
    });

    test('Should pass through the property of HTMLDivElement correctly', () => {
        const wrapper = render(
            <Breadcrumb routes={mockData} title="test" data-jest="test" />
        );
        const container = wrapper.container.querySelector(
            `.${defaultBreadcrumbClassName}`
        ) as HTMLDivElement;

        expect(container?.title).toBe('test');
        expect(container?.dataset.jest).toBe('test');
    });

    test('Should forbid the default of contextMenu', () => {
        const wrapper = render(<Breadcrumb routes={mockData} />);
        const doms = wrapper.container.querySelectorAll(
            `.${breadcrumbItemClassName}`
        );
        const breadcrumb = doms[0];

        const myEvent = createEvent.contextMenu(breadcrumb);
        Object.assign(myEvent, { preventDefault: jest.fn() });
        fireEvent(breadcrumb, myEvent);

        expect(myEvent.preventDefault).toBeCalledTimes(1);
    });
});
