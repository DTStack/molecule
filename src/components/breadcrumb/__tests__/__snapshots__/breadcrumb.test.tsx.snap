// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Breadcrumb Component Should render Breadcrumb multiply levels 1`] = `
<div
  className="mo-breadcrumb"
>
  <a
    className="mo-breadcrumb__item"
    onClick={[Function]}
    onContextMenu={[Function]}
    tabIndex={0}
  >
    <span
      className="mo-breadcrumb__label"
    >
      name0
    </span>
    <span
      className="codicon codicon-chevron-right"
    />
  </a>
  <a
    className="mo-breadcrumb__item"
    onClick={[Function]}
    onContextMenu={[Function]}
    tabIndex={0}
  >
    <span
      className="mo-breadcrumb__label"
    >
      name1
    </span>
    <span
      className="codicon codicon-chevron-right"
    />
  </a>
  <a
    className="mo-breadcrumb__item"
    onClick={[Function]}
    onContextMenu={[Function]}
    tabIndex={0}
  >
    <span
      className="mo-breadcrumb__label"
    >
      name2
    </span>
  </a>
</div>
`;
